### ISSUE TITLE
[Feature Request] nn.Log (or improvement of torch.log)

### ISSUE THREAD

**Issue by rishida0141** [pytorch/pytorch#1620]
I want to add the output of attention layer to the output of decoding layer like Equation(9) of this paper(https://arxiv.org/abs/1704.04368).

This model is called as Pointer-Networks.
The loss function of this model is '- torch.log(p * nn.Softmax(x1) + (1 - p) * nn.Softmax(x2))'.
(nn.LogSoftmax class is not available for this calculation.)

However, there is a difference between the behavior of nn.LogSoftmax(x) and that of torch.log(nn.Softmax(x)).
Because of that, the gradient sometimes get "nan".

Someone please implement nn.Log class to avoid this problem.

This is a sample code to confirm 'nan' gradient.
```
import torch
from torch import nn
from torch import autograd
from torch import optim
import numpy

sm = nn.Softmax()
lsm = nn.LogSoftmax()

crit = nn.NLLLoss()
target = data = torch.LongTensor([1])
target = torch.autograd.Variable(target)

data = torch.FloatTensor([[0, 50, -50]])
data = torch.autograd.Variable(data, requires_grad=True, volatile=False)

num_loop = 1000

for i in xrange(num_loop):
    model = nn.Linear(3, 3, bias=False)
    optimizer = optim.SGD(model.parameters(), lr = 0.01)
    optimizer.zero_grad()

    loss_sm = crit(torch.log(sm(model(data))), target)
    loss_lsm = crit(lsm(model(data)), target)

    # When I use nn.LogSoftmax() class, all gradients are calculated correctly.
    #loss_lsm.backward()
    loss_sm.backward()

    optimizer.step()
    tmp_grad = data.grad.data.cpu().numpy()
    if numpy.isnan(tmp_grad).any():
        print "gradient\n", tmp_grad
        print "parameter\n", model._parameters
        print "gradients contain 'nan'"
        break
if i == (num_loop - 1):
    print "All gradients were calculated correctly"
    print "parameter\n", model._parameters
```

The output of above code
```
gradient
[[ nan  nan  nan]]
parameter
OrderedDict([('weight', Parameter containing:
nan -inf inf
nan nan nan
nan -inf inf
[torch.FloatTensor of size 3x3]
), ('bias', None)])
gradients contain 'nan'
```

**Comment by Jiaming-Liu** [comment]
A relative issue: https://github.com/pytorch/pytorch/issues/1516

You might temporary define a Function with your own backward code.

**Comment by rishida0141** [comment]
Thank you for your comment.

I understood the reason why this problem happen.
The "nan" of the gradient is caused by the backward function of torch.log.

The gradient is dividied by the input of torch.log in the backward function.
If one of the input is quite small, the gradient explode.
(When the number of the class is large, this phenomenon sometimes happen.)

I use torch.masked_select to avoid this problem.

If torch.masked_select is used before torch.log, 
the gradient is divided by the specific value of the input of torch.log.
If the specific value is large enough ( > 10 ** (-30)), the gradient never explode.

This is a sample code
```
import torch
from torch import nn
from torch import autograd
from torch import optim
import numpy

sm = nn.Softmax()

crit = nn.NLLLoss()

data = torch.FloatTensor([[0, 50, -50], [100, 0, -100]])
data = torch.autograd.Variable(data, requires_grad=True, volatile=False)

num_class = 5

target = torch.LongTensor([3, 1])
target = torch.autograd.Variable(target)

mask = [[1 if k == j else 0 for k in range(num_class)] for j in target.data]

num_loop = 1000

for i in xrange(num_loop):

    # This model does not have enough layer to learn correctly.
    model = nn.Linear(3, num_class)

    #model = nn.Sequential(
    #    nn.Linear(3, 2 *num_class),
    #    nn.Tanh(),
    #    nn.Linear(2 * num_class, num_class)
    #)

    optimizer = optim.SGD(model.parameters(), lr = 0.01)
    optimizer.zero_grad()
    
    output = model(data)
    output = output - float(torch.max(output).data.cpu().numpy())
    p = sm(output)

    tmp_mask = autograd.Variable(torch.ByteTensor(mask))
    p_masked = torch.masked_select(p, tmp_mask)
    p_masked += (p_masked < 10 ** (-30)).float() * (10 ** (-30))
    ll = torch.log(p_masked).unsqueeze(1)
    
    loss_sm = crit(ll, autograd.Variable(torch.LongTensor([0, 0])))
    loss_sm.backward()
        
    optimizer.step()
    tmp_grad = data.grad.data.cpu().numpy()
    if numpy.isnan(tmp_grad).any():
        print "gradients contain 'nan'"
        print "gradient\n", tmp_grad
        break
        
if i == (num_loop - 1):
    print "All gradients were calculated correctly"

print "parameter\n", model._parameters
print "prob\n", p

```

**Comment by apaszke** [comment]
Note that a way simpler than mask_select would be to do `p.clamp(min=1e-8)`. This will ensure that all small values won't explode to infinity in backward.

**Comment by rishida0141** [comment]
Thank you for giving me a good advice!
torch.clamp is a good solution to the problem.
I will use this.

**Comment by VictorZuanazzi** [comment]
The suggestion of @apaszke works really well for me. I am training GANs:

`z = torch.randn(current_batch, latent_dim, device = device)
l_G = -torch.log(discriminator(generator(z))).sum()
l_G.clamp(epsilon)
optimizer_G.zero_grad()
l_G.backward()
optimizer_G.step()`
